test/helper/sending.bash

```diff
@@ -82,20 +155,50 @@ function _send_email() {
 # chosen. Sending more than one mail at any given point in time with this function
 # is UNDEFINED BEHAVIOR!
 function _send_email_and_get_id() {
-  [[ -v CONTAINER_NAME ]] || return 1
+  # Get the name of the variable that the ID is stored in
+  local ID_ENV_VAR_NAME="${1:?Mail ID must be set for _send_email_and_get_id}"
+  # Get a "reference" to the content of ID_ENV_VAR_NAME so we can manipulate the content
+  local -n ID_ENV_VAR_REF=${ID_ENV_VAR_NAME}
+  # Export the variable so everyone has access
+  # shellcheck disable=SC2163
+  export "${ID_ENV_VAR_NAME}"
+  shift 1
 
-  _wait_for_empty_mail_queue_in_container
-  _send_email "${@}"
-  _wait_for_empty_mail_queue_in_container
-
-  local MAIL_ID
+  local QUEUE_ID MESSAGE_ID
   # The unique ID Postfix (and other services) use may be different in length
   # on different systems (e.g. amd64 (11) vs aarch64 (10)). Hence, we use a
   # range to safely capture it.
-  MAIL_ID=$(_exec_in_container tac /var/log/mail.log              \
-    | grep -E -m 1 'postfix/smtpd.*: [A-Z0-9]+: client=localhost' \
-    | grep -E -o '[A-Z0-9]{9,12}' || true)
+  local QUEUE_ID_REGEX='[A-Z0-9]{9,12}'
+  local MESSAGE_ID_REGEX="[0-9]{14}\\.${QUEUE_ID_REGEX}"
+
+  _wait_for_empty_mail_queue_in_container
+  if [[ ${1} == --expect-rejection ]]; then
+    shift 1
+    local OUTPUT=$(_send_email_unchecked "${@}")
+    # Because we expect the mail to be rejected, we have to query the mail log
+    # instead of `swaks`, because `swaks` cannot provide us with a queue ID when
+    # mail is rejected (we see something like this instead: `<** 554 5.7.1 Gtube pattern`).
+    QUEUE_ID=$(_exec_in_container tac /var/log/mail.log       \
+      | grep -E "postfix/smtpd.*: ${QUEUE_ID_REGEX}: client=" \
+      | grep -E -m 1 -o '[A-Z0-9]{9,12}' || :)
+  else
+    local OUTPUT=$(_send_email "${@}")
+    # When mail is expected to be delivered, we can use the output of `swaks`
+    # to easily query the queue ID.
+    QUEUE_ID=$(grep -F 'queued as' <<< "${OUTPUT}" | grep -E -o "${QUEUE_ID_REGEX}$")
+  fi
+  _wait_for_empty_mail_queue_in_container
+
+  assert_not_equal "${QUEUE_ID}" ''
+
+  MESSAGE_ID=$(_exec_in_container tac /var/log/mail.log \
+    | grep -E "message-id=<${MESSAGE_ID_REGEX}@"        \
+    | grep -E -m 1 -o "${MESSAGE_ID_REGEX}" || :)
+
+  ID_ENV_VAR_REF="${QUEUE_ID}|${MESSAGE_ID}"
```

Comment:
[Since context for future readers is missing](https://github.com/docker-mailserver/docker-mailserver/pull/3747#discussion_r1443874349). This isn't serving any specific purpose right now, and was included for potential value in future right?

Perhaps it would be good to add a comment in to clarify that intention to future readers that would otherwise see this as dead-code? (_which it technically will be if the suggestion I've provided to add a custom message id header is used_)

Reply:
Oh, I think there is a grave misunderstanding here. Let me explain: this is not dead code but dead important :D `ID_ENV_VAR_NAME` stores _the name_ of the environment variable the caller provided, `ID_ENV_VAR_REF` stores _a reference_ to the contents of the ENV variable the caller provided.

Let's do an example. Say we call `_send_email_and_get_id MY_NICE_ENV`. Now, let's look at what `_send_email_and_get_id` is doing

1. It creates the variable `ID_ENV_VAR_NAME` local to this function

    https://github.com/docker-mailserver/docker-mailserver/blob/c78b6d18475cd82dd07c11fa9bbfe846db06f4b8/test/helper/sending.bash#L133

   which has the content `MY_NICE_ENV`.
2. Then, it creates another variable `ID_ENV_VAR_REF` (**but** with `local -n`)

   https://github.com/docker-mailserver/docker-mailserver/blob/c78b6d18475cd82dd07c11fa9bbfe846db06f4b8/test/helper/sending.bash#L135

    which now has **no contents**! Think of `ID_ENV_VAR_REF` as

    ```rust
    fn _send_email_and_get_id(
        ID_ENV_VAR_NAME: &'static str,
        ID_ENV_VAR_REF: &mut std::env::my_variable_type) { ... }

    // later, we call with
    let MY_NICE_ENV: std::env::my_variable_type;
    _send_email_and_get_id("MY_NICE_ENV", &mut MY_NICE_ENV); 
    ```
3. What happens now I cannot produce in Rust (which is a good thing really, mainly because Rust has sane scoping, unlike Bash...): we `export` the variable `MY_NICE_ENV` (to make it available to everyone) with

    https://github.com/docker-mailserver/docker-mailserver/blob/c78b6d18475cd82dd07c11fa9bbfe846db06f4b8/test/helper/sending.bash#L138

    Only at this point has `MY_NICE_ENV` become an environment variable (which is visible to the caller too). Remember: the caller does not create the variable, it just provides the name of the variable the caller wants the ID to be stored in. The export is the sole reason we want `ID_ENV_VAR_NAME`, because I think it provides better readability than

    ```bash
    function _send_email_and_get_id() {
      export "${1:?}"
      local -n ID_ENV_VAR_REF=${1:?}
    ```

    but I guess it's still Bash, no matter the angle from which you look at it.
4. Later, we store the actual contents with

    https://github.com/docker-mailserver/docker-mailserver/blob/c78b6d18475cd82dd07c11fa9bbfe846db06f4b8/test/helper/sending.bash#L171

---

That being said, I realized this can be done in a more straight-forward manner:

```bash
function _send_email_and_get_id() {
  # Export the variable denotes by ${1} so everyone has access
  export "${1:?Mail ID must be set for _send_email_and_get_id}"
  # Get a "reference" to the content of the variable denoted by ${1} so we can manipulate the content
  local -n ID_ENV_VAR_REF=${1:?}
  # Get rid of ${1} so only the arguments for swaks remain
  shift 1

  ...
```

It's still bash, but potentially easier to grok...

Reply:
> Oh, I think there is a grave misunderstanding here. Let me explain: this is not dead code but dead important :D

Sorry the review comment added that variable in, I didn't consider that dead-code and understood it from my similar implementation of externally referenced vars in `setup.bash` for starting containers with extra args :)

I was actually referring to `MESSAGE_ID` as dead-code. We only have an interest in and use the Queue ID right?

Reply:
Ah, I see. Yes :) It's already removed with the latest commit 🚀